### PR TITLE
Kafka 12766 - Disabling WAL-related Options in RocksDB

### PR DESCRIPTION
**Pull request by TomerWizman**
**description**
Streams disables the write-ahead log (WAL) provided by RocksDB since it replicates the data in changelog topics. Hence, it does not make much sense to set WAL-related configs for RocksDB.
 
**Proposed solution**
Ignore any WAL-related configuration and state in the log that we are ignoring them.  

**Implementation details**
See below
In addition - I made very small refactoring - class RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter is now package private, as this class is internal to Streams

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java [hunk 1]
```diff
@@ -61,12 +62,13 @@
  *
  * This class do the translation between generic {@link Options} into {@link DBOptions} and {@link ColumnFamilyOptions}.
  */
-public class RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter extends Options {
+class RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter extends Options {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.class);
```
Comment (abbccdda) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java]: Use `log` as the name of Logger instance here to be consistent with logger usage in general.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java [hunk 2]
```diff
@@ -304,7 +306,7 @@ public String walDir() {
 
     @Override
     public Options setWalDir(final String walDir) {
-        dbOptions.setWalDir(walDir);
+        LOGGER.warn("WAL is explicitly disabled by Streams in RocksDB. Setting option 'walDir' will be ignored");
```
Comment (abbccdda) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java]: We could refactor out a helper for the log-and-return behavior with parameter as the name of `option`

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java [hunk 3]
```diff
@@ -178,15 +178,15 @@ void openDB(final Map<String, Object> configs, final File stateDir) {
...
+        setStatisticsIfNeeded(configs);
 
         openRocksDB(dbOptions, columnFamilyOptions);
         open = true;
 
         addValueProvidersToMetricsRecorder();
     }
 
-    private void maybeSetUpStatistics(final Map<String, Object> configs) {
+    private void setStatisticsIfNeeded(final Map<String, Object> configs) {
```
Comment (abbccdda) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java]: I don't see why this change is required.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java (hunk 2 above)
Comment (TomerWizman) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java]: sounds like a good suggestion. I can extract the logging to a helper method. I think the return behaviour should still remain under the responsibility of the setters. Let me know what you think

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java (hunk 3 above)
Comment (TomerWizman) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java]: I saw that the "IfNeeded" convention is used in a lot of places. I can revert it

Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java]: `maybeDoSomething()` is also used in a lot of places. So I agree with @abbccdda.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java (hunk 2 above)
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java]: A logging helper method sounds reasonable. IMO the return can stay as it is, because we would not save too much. But I do not have strong feelings about it.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java [hunk 4]
```diff
@@ -249,7 +251,7 @@ public int maxFileOpeningThreads() {
 
     @Override
     public Options setMaxTotalWalSize(final long maxTotalWalSize) {
-        dbOptions.setMaxTotalWalSize(maxTotalWalSize);
+        LOGGER.warn("WAL is explicitly disabled by Streams in RocksDB. Setting option 'maxTotalWalSize' will be ignored");
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java]: Could you please add tests that verify the log messages? You can find an example how to verify log messages in `KTableSourceTest#kTableShouldLogAndMeterOnSkippedRecords()`.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java [hunk 5]
```diff
@@ -61,12 +62,13 @@
  *
  * This class do the translation between generic {@link Options} into {@link DBOptions} and {@link ColumnFamilyOptions}.
  */
-public class RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter extends Options {
+class RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter extends Options {
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java]: I think this is not strictly needed since the constructor is already package-private.

**Review by cadonna — COMMENTED** [review]
Thank you for the PR, @TomerWizman !

I left some comments.

In addition, you have some related test failures in the builds that you need to fix.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java (hunk 1 above)
Comment (TomerWizman) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java]: @abbccdda done

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java (hunk 3 above)
Comment (TomerWizman) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java]: @abbccdda @cadonna I reverted the name back to what it was before.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java (hunk 4 above)
Comment (TomerWizman) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java]: @cadonna done. let me know if this is what you wanted

Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java]: Yes, it is! Could add also the other ignored WAL options to the test?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java [hunk 6]
```diff
@@ -37,7 +37,7 @@
 import org.rocksdb.DbPath;
 import org.rocksdb.Env;
 import org.rocksdb.InfoLogLevel;
-import org.rocksdb.Logger;
+
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java]:
```suggestion
```

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java (hunk 4 above)
Comment (TomerWizman) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.java]: @cadonna 
Sorry for the late reply, was on vacation.
I added the rest of the wal options to the test.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java [hunk 7]
```diff
@@ -93,7 +114,7 @@ public void shouldOverwriteAllOptionsMethods() throws Exception {
         for (final Method method : Options.class.getMethods()) {
             if (!ignoreMethods.contains(method.getName())) {
                 RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.class
-                    .getDeclaredMethod(method.getName(), method.getParameterTypes());
+                        .getDeclaredMethod(method.getName(), method.getParameterTypes());
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java]: Could you please revert this change since it is unrelated to the PR?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java [hunk 8]
```diff
@@ -113,7 +134,7 @@ public void shouldForwardAllDbOptionsCalls() throws Exception {
     private void verifyDBOptionsMethodCall(final Method method) throws Exception {
         final DBOptions mockedDbOptions = mock(DBOptions.class);
         final RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter optionsFacadeDbOptions
-            = new RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter(mockedDbOptions, new ColumnFamilyOptions());
+                = new RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter(mockedDbOptions, new ColumnFamilyOptions());
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java]: Could you please revert this change since it is unrelated to the PR?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java [hunk 9]
```diff
@@ -126,7 +147,7 @@ private void verifyDBOptionsMethodCall(final Method method) throws Exception {
         } catch (final InvocationTargetException undeclaredMockMethodCall) {
             assertThat(undeclaredMockMethodCall.getCause(), instanceOf(AssertionError.class));
             assertThat(undeclaredMockMethodCall.getCause().getMessage().trim(),
-                matchesPattern("Unexpected method call DBOptions\\." + method.getName() + "((.*\n*)*):"));
+                    matchesPattern("Unexpected method call DBOptions\\." + method.getName() + "((.*\n*)*):"));
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java]: Could you please revert this change since it is unrelated to the PR?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java [hunk 10]
```diff
@@ -218,7 +239,7 @@ public void shouldForwardAllColumnFamilyCalls() throws Exception {
     private void verifyColumnFamilyOptionsMethodCall(final Method method) throws Exception {
         final ColumnFamilyOptions mockedColumnFamilyOptions = mock(ColumnFamilyOptions.class);
         final RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter optionsFacadeColumnFamilyOptions
-            = new RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter(new DBOptions(), mockedColumnFamilyOptions);
+                = new RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter(new DBOptions(), mockedColumnFamilyOptions);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java]: Could you please revert this change since it is unrelated to the PR?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java [hunk 11]
```diff
@@ -231,7 +252,7 @@ private void verifyColumnFamilyOptionsMethodCall(final Method method) throws Exc
         } catch (final InvocationTargetException undeclaredMockMethodCall) {
             assertThat(undeclaredMockMethodCall.getCause(), instanceOf(AssertionError.class));
             assertThat(undeclaredMockMethodCall.getCause().getMessage().trim(),
-                matchesPattern("Unexpected method call ColumnFamilyOptions\\." + method.getName() +  "(.*)"));
+                    matchesPattern("Unexpected method call ColumnFamilyOptions\\." + method.getName() +  "(.*)"));
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java]: Could you please revert this change since it is unrelated to the PR?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java [hunk 12]
```diff
@@ -306,4 +327,33 @@ public String name() {
...
         return parameters;
     }
+
+    @Test
+    public void shouldLogWarningWhenSettingWalOptions() throws Exception {
+
+        try (final LogCaptureAppender appender = LogCaptureAppender.createAndRegister(RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.class)) {
+
+            final RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter adapter
+                    = new RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter(new DBOptions(), new ColumnFamilyOptions());
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java]: nit:
```suggestion
                = new RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter(new DBOptions(), new ColumnFamilyOptions());
```

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java [hunk 13]
```diff
@@ -306,4 +327,33 @@ public String name() {
...
+                    method.invoke(adapter, getDBOptionsParameters(method.getParameterTypes()));
+                }
+            }
+
+            final List<String> walOptions = Arrays.asList("walDir", "walFilter", "walRecoveryMode", "walBytesPerSync", "walSizeLimitMB", "manualWalFlush", "maxTotalWalSize", "walTtlSeconds");
+
+            final Set<String> logMessages = appender.getEvents().stream()
+                    .filter(e -> e.getLevel().equals("WARN"))
+                    .map(LogCaptureAppender.Event::getMessage)
+                    .collect(Collectors.toSet());
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java]: nit:
```suggestion
            final Set<String> logMessages = appender.getEvents().stream()
                .filter(e -> e.getLevel().equals("WARN"))
                .map(LogCaptureAppender.Event::getMessage)
                .collect(Collectors.toSet());
```

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java [hunk 14]
```diff
@@ -306,4 +327,33 @@ public String name() {
...
+            final RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter adapter
+                    = new RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter(new DBOptions(), new ColumnFamilyOptions());
+
+            for (final Method method : RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter.class.getDeclaredMethods()) {
+                if (walRelatedMethods.contains(method.getName())) {
+                    method.invoke(adapter, getDBOptionsParameters(method.getParameterTypes()));
+                }
+            }
+
+            final List<String> walOptions = Arrays.asList("walDir", "walFilter", "walRecoveryMode", "walBytesPerSync", "walSizeLimitMB", "manualWalFlush", "maxTotalWalSize", "walTtlSeconds");
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java:345]: nit: Would it be possible to extract this options from `walRelatedMethods`?

**Review by cadonna — COMMENTED** [review]
Thank you for the update, @TomerWizman !

LGTM!

I had just some minor comments.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java (hunk 7 above)
Comment (TomerWizman) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java]: @cadonna 
done, no idea how it came in...

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java (hunk 8 above)
Comment (TomerWizman) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java]: @cadonna 
done, no idea how it came in...

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java (hunk 9 above)
Comment (TomerWizman) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java]: @cadonna 
done, no idea how it came in...

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java (hunk 14 above)
Comment (TomerWizman) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java:345]: @cadonna 
I Thought about it also.
To do that I need to remove the "set" from the method name and to de-capitalize the first letter of the option. 
I think this is a bit overkill because:
1. I got checkstyle errors trying to use something like toLowerCase()... 
2. Its going to be ugly to do it without additional StringUtils helper libs (like apache commons-lang StringUtils which I saw we dont have)

Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapterTest.java:345]: Fair enough!

**Review by cadonna — APPROVED** [review]
LGTM!

Thank you, @TomerWizman !

**Comment by cadonna** [comment]
Build failures are unrelated.
